pilot/pkg/xds/proxy_dependencies.go

```diff
@@ -35,6 +35,13 @@ func ConfigAffectsProxy(req *model.PushRequest, proxy *model.Proxy) bool {
 		return true
 	}
 
+	// When endpoints for headless service change, we should push if EnableHeadlessService is true
+	// so that we create updated listeners or when DNSCapture is enabled for proxy so that the NDS
+	// table is updated.
+	if req.Full && headlessEndpointsUpdated(req) && (features.EnableHeadlessService || bool(proxy.Metadata.DNSCapture)) {
```

Comment:
This may introduce additional full push when  EnableHeadlessService = false, actually in this case we only need NDS and EDS?

Reply:
How about updating `ndsNeedsPush`,  push nds on headless service endpoint updated

Reply:
The diff is make full = false and you can set reason = HeadlessEndpoint.

Reply:
I think we can make full = false because it is needed when EnableHeadlessService is enabled. But changed a bit - hopefully this is better. PTAL